### PR TITLE
chore: upgrade kube-openapi Fixes #9149

### LINKED ISSUE #9149: CRITICAL Vulnerability CVE-2022-1996 found in image for v3.3.8
The docker image for argoproj/argo-workflows:v3.3.8 (https://github.com/argoproj/argo-workflows/releases/tag/v3.3.8) is flagging vulnerability [CVE-2022-1996 ](https://nvd.nist.gov/vuln/detail/CVE-2022-1996). 

The vulnerability appears to come from the underlying argo docker images (argoexec, argocli, workflow-container). The :latest images for these (in docker and quay) no longer have the vulnerability. Can you please rebuild the image to use the latest images and release a new version?

### PR DESCRIPTION
**Pull request by mcgrawia**
Signed-off-by: Ian McGraw <mcgrawia@umich.edu>

Fixes #9149 CVE-2022-1996

### CODE REVIEW

File: v3 [hunk 1]
```diff
@@ -0,0 +1 @@
+.
```
Comment (terrytangyuan) [inline v3:1]: Could you remove this file?

Comment (mcgrawia) [reply v3:1]: hi @terrytangyuan, the v3 file came from the output of the `make pre-commit -B` command the pull request instructions said to run first. I can remove though 👍

File: test/util/shared_index_informer.go [hunk 2]
```diff
@@ -26,3 +27,4 @@ func (s *SharedIndexInformer) LastSyncResourceVersion() string
 func (s *SharedIndexInformer) AddIndexers(cache.Indexers) error                           { return nil }
 func (s *SharedIndexInformer) GetIndexer() cache.Indexer                                  { return s.Indexer }
 func (s *SharedIndexInformer) SetWatchErrorHandler(handler cache.WatchErrorHandler) error { return nil }
+func (s *SharedIndexInformer) SetTransform(handler cache.TransformFunc) error             { panic("implement me") }
```
Comment (mcgrawia) [inline test/util/shared_index_informer.go:30]: this was a new function added to the `SharedIndexInformer` interface in k8s client-go v0.24.3

**Comment by mcgrawia** [comment]
@terrytangyuan can I get some help on how to resolve the Codegen issue? I attempted to run the command in the pull request template `make pre-commit -B`, but it doesn't seem to have changed anything except making the empty `v3` file you had me remove above. Thanks

**Comment by terrytangyuan** [comment]
When you ran the command, did you see any errors?

**Comment by mcgrawia** [comment]
I didn't realize it the first time but it looks like there are some, although I'm not sure what the fix should be. Here's the full log: 
```
 make pre-commit -B                                                                              
GIT_COMMIT=12f55a6bde9a5090ae9101e5ebdbee371dc0f752 GIT_BRANCH=fix-CVE-2022-1996 GIT_TAG=untagged GIT_TREE_STATE=clean RELEASE_TAG=false DEV_BRANCH=true VERSION=latest
KUBECTX=docker-desktop DOCKER_DESKTOP=true K3D=false DOCKER_PUSH=false
RUN_MODE=local PROFILE=minimal AUTH_MODE=hybrid SECURE=false STATIC_FILES=false ALWAYS_OFFLOAD_NODE_STATUS=false UPPERIO_DB_DEBUG=0 LOG_LEVEL=debug NAMESPACED=true AZURE=false
⚠️ Code generation will not work if code in not checked out into /Users/ianmcgraw/go/src/github.com/argoproj/argo-workflows
go install k8s.io/code-generator/cmd/go-to-protobuf@v0.21.5
go install github.com/gogo/protobuf/protoc-gen-gogo@v1.3.2
go install github.com/gogo/protobuf/protoc-gen-gogofast@v1.3.2
go install golang.org/x/tools/cmd/goimports@v0.1.7
go install github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway@v1.16.0
go install github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger@v1.16.0
brew install clang-format
Running `brew update --auto-update`...
==> Auto-updated Homebrew!
Updated 4 taps (aquasecurity/trivy, homebrew/cask-versions, homebrew/core and homebrew/cask).
==> New Formulae
dura
==> New Casks
polypad

You have 70 outdated formulae and 1 outdated cask installed.
You can upgrade them with brew upgrade
or list them with brew outdated.

Warning: clang-format 14.0.6 is already installed and up-to-date.
To reinstall 14.0.6, run:
  brew reinstall clang-format
[ -e /Users/ianmcgraw/go/src/github.com/gogo/protobuf ] || git clone --depth 1 https://github.com/gogo/protobuf.git -b v1.3.2 /Users/ianmcgraw/go/src/github.com/gogo/protobuf
# These files are generated on a v3/ folder by the tool. Link them to the root folder
[ -e ./v3 ] || ln -s . v3
# Format proto files. Formatting changes generated code, so we do it here, rather that at lint time.
# Why clang-format? Google uses it.
find pkg/apiclient -name '*.proto'|xargs clang-format -i
/Users/ianmcgraw/go/bin/go-to-protobuf \
                --go-header-file=./hack/custom-boilerplate.go.txt \
                --packages=github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1 \
                --apimachinery-packages=+k8s.io/apimachinery/pkg/util/intstr,+k8s.io/apimachinery/pkg/api/resource,k8s.io/apimachinery/pkg/runtime/schema,+k8s.io/apimachinery/pkg/runtime,k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/api/core/v1,k8s.io/api/policy/v1beta1 \
                --proto-import /Users/ianmcgraw/go/src
topological order bytes
topological order encoding/base64
topological order encoding/json
topological order errors
topological order fmt
topological order github.com/argoproj/argo-workflows/v3/errors
topological order github.com/argoproj/argo-workflows/v3/pkg/apis/workflow
topological order github.com/argoproj/argo-workflows/v3/util/json
topological order github.com/argoproj/argo-workflows/v3/util/slice
topological order github.com/gogo/protobuf/proto
topological order github.com/gogo/protobuf/sortkeys
topological order github.com/google/gofuzz
topological order go/ast
topological order go/doc
topological order go/parser
topological order go/token
topological order gopkg.in/inf.v0
topological order hash/fnv
topological order io
topological order io/ioutil
topological order k8s.io/apimachinery/pkg/conversion
topological order k8s.io/apimachinery/pkg/conversion/queryparams
topological order k8s.io/apimachinery/pkg/fields
topological order k8s.io/apimachinery/pkg/labels
topological order k8s.io/apimachinery/pkg/selection
topological order k8s.io/apimachinery/pkg/types
topological order k8s.io/apimachinery/pkg/util/errors
topological order k8s.io/apimachinery/pkg/util/json
topological order k8s.io/apimachinery/pkg/util/naming
topological order k8s.io/apimachinery/pkg/util/runtime
topological order k8s.io/apimachinery/pkg/util/sets
topological order k8s.io/apimachinery/pkg/util/validation
topological order k8s.io/apimachinery/pkg/util/wait
topological order k8s.io/apimachinery/pkg/watch
topological order k8s.io/klog/v2
topological order k8s.io/kube-openapi/pkg/common
topological order k8s.io/kube-openapi/pkg/validation/spec
topological order math
topological order math/big
topological order math/bits
topological order net/http
topological order net/url
topological order os
topological order path
topological order path/filepath
topological order reflect
topological order regexp
topological order runtime/debug
topological order sigs.k8s.io/structured-merge-diff/v4/value
topological order sigs.k8s.io/yaml
topological order sort
topological order strconv
topological order strings
topological order sync
topological order sync/atomic
topological order time
topological order unsafe
topological order k8s.io/apimachinery/pkg/runtime/schema
topological order k8s.io/apimachinery/pkg/util/intstr
topological order k8s.io/apimachinery/pkg/api/resource
topological order k8s.io/apimachinery/pkg/runtime
topological order k8s.io/apimachinery/pkg/apis/meta/v1
topological order k8s.io/api/core/v1
topological order k8s.io/api/policy/v1beta1
topological order github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1
2022/07/27 10:10:33 /Users/ianmcgraw/dev/argo-workflows/vendor/k8s.io/apimachinery/pkg/util/intstr/generated.proto: File does not reside within any path specified using --proto_path (or -I).  You must specify a --proto_path which encompasses this file.  Note that the proto_path must be an exact prefix of the .proto file names -- protoc is too dumb to figure out when two paths (e.g. absolute and relative) are equivalent (it's harder than you think).
2022/07/27 10:10:33 protoc -I . -I /Users/ianmcgraw/go/src -I /Users/ianmcgraw/go/src --gogo_out=/Users/ianmcgraw/go/src /Users/ianmcgraw/dev/argo-workflows/vendor/k8s.io/apimachinery/pkg/util/intstr/generated.proto
2022/07/27 10:10:33 Unable to generate protoc on k8s.io.apimachinery.pkg.util.intstr: exit status 1
make: *** [pkg/apis/workflow/v1alpha1/generated.proto] Error 1
```

**Comment by mcgrawia** [comment]
also when I check out the code according to that warning in `/Users/ianmcgraw/go/src/github.com/argoproj/argo-workflows` I get the following errors:
```
 make pre-commit -B
GIT_COMMIT=12f55a6bde9a5090ae9101e5ebdbee371dc0f752 GIT_BRANCH=fix-CVE-2022-1996 GIT_TAG=untagged GIT_TREE_STATE=clean RELEASE_TAG=false DEV_BRANCH=true VERSION=latest
KUBECTX=docker-desktop DOCKER_DESKTOP=true K3D=false DOCKER_PUSH=false
RUN_MODE=local PROFILE=minimal AUTH_MODE=hybrid SECURE=false STATIC_FILES=false ALWAYS_OFFLOAD_NODE_STATUS=false UPPERIO_DB_DEBUG=0 LOG_LEVEL=debug NAMESPACED=true AZURE=false
⚠️ Code generation will not work if code in not checked out into /src/github.com/argoproj/argo-workflows
go install k8s.io/code-generator/cmd/go-to-protobuf@v0.21.5
go install github.com/gogo/protobuf/protoc-gen-gogo@v1.3.2
go install github.com/gogo/protobuf/protoc-gen-gogofast@v1.3.2
go install golang.org/x/tools/cmd/goimports@v0.1.7
go install github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway@v1.16.0
go install github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger@v1.16.0
brew install clang-format
Running `brew update --auto-update`...
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).

You have 70 outdated formulae and 1 outdated cask installed.
You can upgrade them with brew upgrade
or list them with brew outdated.

Warning: clang-format 14.0.6 is already installed and up-to-date.
To reinstall 14.0.6, run:
  brew reinstall clang-format
[ -e /Users/ianmcgraw/go/src/github.com/gogo/protobuf ] || git clone --depth 1 https://github.com/gogo/protobuf.git -b v1.3.2 /Users/ianmcgraw/go/src/github.com/gogo/protobuf
# These files are generated on a v3/ folder by the tool. Link them to the root folder
[ -e ./v3 ] || ln -s . v3
# Format proto files. Formatting changes generated code, so we do it here, rather that at lint time.
# Why clang-format? Google uses it.
find pkg/apiclient -name '*.proto'|xargs clang-format -i
/Users/ianmcgraw/go/bin/go-to-protobuf \
		--go-header-file=./hack/custom-boilerplate.go.txt \
		--packages=github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1 \
		--apimachinery-packages=+k8s.io/apimachinery/pkg/util/intstr,+k8s.io/apimachinery/pkg/api/resource,k8s.io/apimachinery/pkg/runtime/schema,+k8s.io/apimachinery/pkg/runtime,k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/api/core/v1,k8s.io/api/policy/v1beta1 \
		--proto-import /Users/ianmcgraw/go/src
topological order bytes
topological order encoding/base64
topological order encoding/json
topological order errors
topological order fmt
topological order github.com/argoproj/argo-workflows/v3/errors
topological order github.com/argoproj/argo-workflows/v3/pkg/apis/workflow
topological order github.com/argoproj/argo-workflows/v3/util/json
topological order github.com/argoproj/argo-workflows/v3/util/slice
topological order github.com/gogo/protobuf/proto
topological order github.com/gogo/protobuf/sortkeys
topological order github.com/google/gofuzz
topological order go/ast
topological order go/doc
topological order go/parser
topological order go/token
topological order gopkg.in/inf.v0
topological order hash/fnv
topological order io
topological order io/ioutil
topological order k8s.io/apimachinery/pkg/conversion
topological order k8s.io/apimachinery/pkg/conversion/queryparams
topological order k8s.io/apimachinery/pkg/fields
topological order k8s.io/apimachinery/pkg/labels
topological order k8s.io/apimachinery/pkg/selection
topological order k8s.io/apimachinery/pkg/types
topological order k8s.io/apimachinery/pkg/util/errors
topological order k8s.io/apimachinery/pkg/util/json
topological order k8s.io/apimachinery/pkg/util/naming
topological order k8s.io/apimachinery/pkg/util/runtime
topological order k8s.io/apimachinery/pkg/util/sets
topological order k8s.io/apimachinery/pkg/util/validation
topological order k8s.io/apimachinery/pkg/util/wait
topological order k8s.io/apimachinery/pkg/watch
topological order k8s.io/klog/v2
topological order k8s.io/kube-openapi/pkg/common
topological order k8s.io/kube-openapi/pkg/validation/spec
topological order math
topological order math/big
topological order math/bits
topological order net/http
topological order net/url
topological order os
topological order path
topological order path/filepath
topological order reflect
topological order regexp
topological order runtime/debug
topological order sigs.k8s.io/structured-merge-diff/v4/value
topological order sigs.k8s.io/yaml
topological order sort
topological order strconv
topological order strings
topological order sync
topological order sync/atomic
topological order time
topological order unsafe
topological order k8s.io/apimachinery/pkg/runtime/schema
topological order k8s.io/apimachinery/pkg/util/intstr
topological order k8s.io/apimachinery/pkg/api/resource
topological order k8s.io/apimachinery/pkg/runtime
topological order k8s.io/apimachinery/pkg/apis/meta/v1
topological order k8s.io/api/core/v1
topological order k8s.io/api/policy/v1beta1
topological order github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1
2022/07/27 10:23:39 protoc-gen-gogo: program not found or is not executable
Please specify a program using absolute path or make sure the program is available in your PATH system variable
--gogo_out: protoc-gen-gogo: Plugin failed with status code 1.
2022/07/27 10:23:39 protoc -I . -I /Users/ianmcgraw/go/src -I /Users/ianmcgraw/go/src --gogo_out=/Users/ianmcgraw/go/src /Users/ianmcgraw/go/src/github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/util/intstr/generated.proto
2022/07/27 10:23:39 Unable to generate protoc on k8s.io.apimachinery.pkg.util.intstr: exit status 1
make: *** [pkg/apis/workflow/v1alpha1/generated.proto] Error 1
```

File: .github/workflows/ci-build.yaml [hunk 3]
```diff
@@ -17,7 +17,7 @@ jobs:
   tests:
     name: Unit Tests
     runs-on: ubuntu-latest
-    timeout-minutes: 8
```
Comment (sarabala1979) [inline .github/workflows/ci-build.yaml:20]: Sync with the master. Master has unittest improvement

Comment (mcgrawia) [reply .github/workflows/ci-build.yaml:20]: Hi @sarabala1979, my PR is already based on the latest master: 
image

Comment (mcgrawia) [reply .github/workflows/ci-build.yaml:20]: apologies, there was a new commit as of 4 hours ago I did not have, but I don't think this should affect the unittests. Rebased 👍 
https://github.com/mcgrawia/argo-workflows/commit/b44981f174acebac239e4ecbdda845aa4338f5af

**Comment by terrytangyuan** [comment]
`protoc-gen-gogo: program not found or is not executable` - looks like this binary isn't installed. Perhaps try `go install github.com/gogo/protobuf/protoc-gen-gogo@v1.3.2` and then run again?

**Comment by mcgrawia** [comment]
@terrytangyuan when I run that command, no output occurs, so I think it's already installed:
```shell
go install github.com/gogo/protobuf/protoc-gen-gogo@v1.3.2
```
I can see the binary in my `~/go/bin` folder:
```shell
ls ~/go/bin
client-gen	deepcopy-gen   diagram	  go-to-protobuf  goimports  informer-gen  jsonschema  kind	   openapi-gen	    protoc-gen-gogofast      protoc-gen-swagger
controller-gen	defaulter-gen  generator  go1.18.4	  hack	     jose-util	   jwk-keygen  lister-gen  protoc-gen-gogo  protoc-gen-grpc-gateway  swagger
```

**Comment by mcgrawia** [comment]
I wiped my Go installation and reinstalled everything and added `~/go/bin` to my path, now I'm getting the following error:
```
topological order k8s.io/apimachinery/pkg/api/resource
topological order k8s.io/apimachinery/pkg/runtime
topological order k8s.io/apimachinery/pkg/apis/meta/v1
topological order k8s.io/api/core/v1
topological order k8s.io/api/policy/v1beta1
topological order github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1
2022/07/27 16:31:11 k8s.io/apimachinery/pkg/runtime/generated.proto: File not found.
k8s.io/apimachinery/pkg/runtime/schema/generated.proto: File not found.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:9:1: Import "k8s.io/apimachinery/pkg/runtime/generated.proto" was not found or had errors.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:10:1: Import "k8s.io/apimachinery/pkg/runtime/schema/generated.proto" was not found or had errors.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:469:12: "k8s.io.apimachinery.pkg.runtime.RawExtension" is not defined.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:1214:12: "k8s.io.apimachinery.pkg.runtime.RawExtension" is not defined.
2022/07/27 16:31:11 protoc -I . -I /Users/ianmcgraw/go/src -I /Users/ianmcgraw/go/src --gogo_out=/Users/ianmcgraw/go/src /Users/ianmcgraw/go/src/github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto
2022/07/27 16:31:11 Unable to generate protoc on k8s.io.apimachinery.pkg.apis.meta.v1: exit status 1
make: *** [pkg/apis/workflow/v1alpha1/generated.proto] Error 1
```.
Is there documentation about how to run this processes? It seems like I'm missing some basic setup

**Comment by terrytangyuan** [comment]
The errors indicate that your `vendor/k8s.io/apimachinery` may not be updated enough. Could you try remove `/vendor` folder and then run `go mod vendor` to regenerate it before running `make pre-commit` again?

**Comment by mcgrawia** [comment]
hi @terrytangyuan , thanks or the help. Unfortunately after deleting the vendor folder and re-running I still see the following error:
```
...
topological order k8s.io/api/policy/v1beta1
topological order github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1
2022/07/28 09:13:10 k8s.io/apimachinery/pkg/runtime/generated.proto: File not found.
k8s.io/apimachinery/pkg/runtime/schema/generated.proto: File not found.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:9:1: Import "k8s.io/apimachinery/pkg/runtime/generated.proto" was not found or had errors.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:10:1: Import "k8s.io/apimachinery/pkg/runtime/schema/generated.proto" was not found or had errors.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:469:12: "k8s.io.apimachinery.pkg.runtime.RawExtension" is not defined.
github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto:1214:12: "k8s.io.apimachinery.pkg.runtime.RawExtension" is not defined.
2022/07/28 09:13:10 protoc -I . -I /Users/ianmcgraw/go/src -I /Users/ianmcgraw/go/src --gogo_out=/Users/ianmcgraw/go/src /Users/ianmcgraw/go/src/github.com/argoproj/argo-workflows/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto
2022/07/28 09:13:10 Unable to generate protoc on k8s.io.apimachinery.pkg.apis.meta.v1: exit status 1
make: *** [pkg/apis/workflow/v1alpha1/generated.proto] Error 1
```

if it helps, my GO env variables look like this:
```shell
echo $GO111MODULE $GOROOT $GOPATH 
on /usr/local/go /Users/ianmcgraw/go
```

**Comment by mcgrawia** [comment]
@terrytangyuan would it be possible for someone on the Argo side to attempt this command and if it succeeds push the generated files to this branch? I imagine someone on the team likely already has this working so it might be easier than having me sink more hours into troubleshooting on my end.

**Comment by terrytangyuan** [comment]
> @terrytangyuan would it be possible for someone on the Argo side to attempt this command and if it succeeds push the generated files to this branch? I imagine someone on the team likely already has this working so it might be easier than having me sink more hours into troubleshooting on my end.

I just pushed the generated files.

**Comment by mcgrawia** [comment]
thank you @terrytangyuan ! I'll see if I can get the final tests passing and hopefully we can get this merged

**Comment by terrytangyuan** [comment]
The build failures may not be related to your changes.

**Comment by mcgrawia** [comment]
@terrytangyuan ok thank you for the heads up, I'll wait to hear from you for next steps

**Comment by mcgrawia** [comment]
@terrytangyuan it looks like all of the tests are now passing, let me know if there's anything else I can do before we can merge this, thanks!

Also I wanted to ask, is there anything else I can do to help get this released ASAP? Is there a branch I should back-port this patch to? Not sure what would be required to release a v3.3.9.

**Review by terrytangyuan — APPROVED** [review]
LGTM. I'd defer to @sarabala1979 @alexec regarding back-porting this.

File: test/util/shared_index_informer.go (hunk 2 above)
Comment (alexec) [reply test/util/shared_index_informer.go:30]: maybe return nil?

**Review by alexec — APPROVED** [review]

**Comment by mcgrawia** [comment]
thanks @alexec, changed! Will merge once CICD finishes. Any suggestions how I can help get this patched into a `v3.3.9` release?

**Comment by mcgrawia** [comment]
also @alexec @terrytangyuan @sarabala1979 I think everything should be good to go on this PR when one of you is available to merge. Thanks for the help 👍

**Comment by terrytangyuan** [comment]
Thank you! 

@sarabala1979 Do you think we can backport this to 3.3.9 release?

**Comment by mcgrawia** [comment]
Good morning @sarabala1979, please let me know if there's anything I can help with to get a 3.3.9 release with this patch. This vulnerability is preventing us from releasing to our customers so would appreciate if we can get this out quickly. Thank you

**Comment by terrytangyuan** [comment]
https://github.com/argoproj/argo-workflows/issues/9262

**Pull request merged**